test/test.cpp

```diff
@@ -3,10 +3,378 @@ bundle config --local path vendor/bundle
 bundle install
 bundle exec arduino_ci_remote.rb  --skip-compilation
  */
+#include <cassert>
+#include <vector>
 #include "Arduino.h"
 #include "ArduinoUnitTests.h"
 #include "Keypad.h"
+#include "ci/ObservableDataStream.h"
 
-unittest(test) { assertTrue(true); }
+class BitCollector : public DataStreamObserver
+{
+private:
+    byte pinC;
+    byte pinR;
+    char observer[5] = "pin ";
+    bool showData;
+    vector<int> pinLog;
+    GodmodeState *state;
 
+public:
+    BitCollector(byte pinC, byte pinR, bool showData = false)
+        : DataStreamObserver(false, false)
+    {
+        assert(pinC >= 0 && pinC < 10 && pinR >= 0 && pinR < 10);
+        observer[4] = '0' + pinC;
+        this->pinR = pinR;
+        this->pinC = pinC;
+        this->showData = showData;
+        state = GODMODE();
+        state->reset();
+        state->digitalPin[pinC].addObserver(observer, this);
+    }
+
+    ~BitCollector() { state->digitalPin[pinC].removeObserver(observer); }
+
+    virtual void onBit(bool aBit)
+    {
+        if (aBit == LOW)
+        {
+            state->digitalPin[pinR] = LOW;
+        }
+        else
+        {
+            state->digitalPin[pinR] = HIGH;
+        }
+    }
+
+    virtual String observerName() const { return observer; }
+};
+
+// ----------------------------------------------------------------------------
```

Comment:
Typically we try to avoid committing a large chunk of commented-out code. If it doesn't work and you plan to work on it, keep it in a separate branch. In any case, it probably should be removed from here.